---
title: RNA, sRNA and TE differential expression analyses in Zebrafish - DR. Alice
  Godden
output:
  word_document: default
  html_document:
    df_print: paged
  pdf_document: default
editor_options:
  chunk_output_type: console
---


Load and install required packages
```{r}
library(DESeq2)
library(tidyverse)
if (!requireNamespace('BiocManager', quietly = TRUE))
  install.packages('BiocManager')

BiocManager::install('EnhancedVolcano')
BiocManager::install('pheatmap')
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("pcaExplorer")

browseVignettes("pcaExplorer")
library("pcaExplorer")
library(ggplot2)
library(biomaRt)
library(tidyverse)
library(org.Dr.eg.db)

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("org.Dr.eg.db")
if (!require("gplots")) {
  install.packages("gplots", dependencies = TRUE)
  library(gplots)
}
if (!require("RColorBrewer")) {
  install.packages("RColorBrewer", dependencies = TRUE)
  library(RColorBrewer)
}
library(readr)
library(viridis)
library(dplyr)
library(ggpubr)
library(ggVennDiagram)

```


Read in your raw counts data
```{r}
#use for rRNA seq and sub in the right file names 
countData <- read.csv('raw_counts.csv')
head(countData)
```

Make your first column, as row names (genes/gene ids etc)
```{r}
#delete first column of row numbers

df = subset(countData, select = -c( 1) )
rownames(df) <- countData[ , 1]
head(df)
mode(df) # tells you type of df you have
```

Read in the metadata information on your samples. Note that
the samples need to be in the same order in this sheet, as they do in your countData (raw counts) file.
```{r}
colData <- read.csv('metadata.csv')
head(colData)
#make first column of colData into row names
rownames(colData) <- colData [,1]  
head(colData)
colData = subset(colData, select = -c(1) )
head(colData)

```

```{r}
#making sure the row names in colData match column names in df (your countdata)
all(colnames(df) %in% rownames(colData))
# if you get true then this is correct

#to ensure they are in same order
all(colnames(df) == rownames(colData))
# if true then this is correct

```

Making the DESeq (dds) object, if you have one factor, eg treatment v control. If your raw count data is whole integers remove the round command on the first line.
```{r}
dds <- DESeqDataSetFromMatrix(countData=round (df),
                              colData = colData,
                              design = ~ group) 
```

Making the DEseq (dds) object, if you have two factors, eg treatment and organ for multiple comparisons
```{r}
dds <- DESeqDataSetFromMatrix(countData=round (df),
                              colData = colData,
                              design = ~ gender + group) 

#assess for interaction between group and gender
dds <- DESeqDataSetFromMatrix(countData=round (df),
                              colData = colData,
                              design = ~ group + gender + group * gender) 


```

Run DEseq
```{r}
#Run DEseq.
dds<-DESeq(dds)

dds

```


Pre-filtering: removing rows with low gene counts
Keeping rows that have at least 10 reads total. This is because low read counts (less than 10) can skew your DESeq results.
```{r}
keep <- rowSums(counts(dds)) >= 10

dds <- dds [keep,]

dds
```

Use resultsNames to find out what conditions (factors) are being examined. Use relevel to change these, then 
```{r}
#step3 run DESeq
dds <- DESeq (dds)
resultsNames(dds)
dds$Gender <-relevel(dds$Gender, ref = "Female")

#re run to check you have relevelled and the contrasts you have in your results with resultsNames, first variable is the result you are looking at. eg. if it says treatment v control, the differentially expressed genes are resultant of the treatment
dds <- DESeq (dds)
resultsNames(dds)
#Now generate results 
res <- results(dds)
```


Can also modify the conditions in the results with this
```{r}
results(dds, contrast=list(c("group_SEG_vs_NEG", "gender_Male_vs_Female"  )))

```

View results, a summary, significant results only and save the LFC results in a .csv file.
```{r}
res

#Explore results
summary(res)


res0.05 <- results(dds, alpha =0.05)
summary(res0.05)

#to get normalised counts
normalized_counts <- counts(dds, normalized = TRUE)
write.csv(normalized_counts, file="normalised_counts.csv")


write.csv(res, file="results.csv")
```

Let's make some plots for analyses
```{r}
# MA plot
plotMA(res)

```



Let's make a PCA plot using PCA explorer, an interactive tool that opens a pop up
```{r}
pcaExplorer(dds = dds)
```


Let's make a PCA plot manually
```{r}

nsub=nrow(dds)

#If you have a big dataset use vst as this can be quicker
rld <- rlog(dds, blind=FALSE)

#for looking at the variance in the top 300 genes on one factor
plotPCA(rld, intgroup = "group", ntop = 300, returnData = TRUE)
pcaData <- plotPCA(rld, intgroup=c("group"), returnData=TRUE)

#for looking at two factors with shape and colour for multiple factors
pcaData <- plotPCA(rld, intgroup=c("group", "age"), returnData=TRUE)
percentVar <- round(300 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=group)) +
  geom_point(size=4, alpha = 0.5) + # alpha for more transparency use a lower number
  scale_color_manual(values = c("red", "blue", "orange")) +
  scale_shape_manual(values=seq(0,15)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  stat_ellipse() +
  coord_fixed()

#+
geom_label(aes(label = name))

z <- plotPCA(rld)
z + geom_label(aes(label = name))

#the PCA's used in this manuscript

rld <- rlog(dds, blind=FALSE)
plotPCA(rld, intgroup = "group", ntop = 300, returnData = TRUE)
pcaData <- plotPCA(rld, intgroup=c("group"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))


# pretty
  p <- ggplot(pcaData, aes(PC1, PC2, color=group)) +
  geom_point(size=6, alpha = 0.2) +
  scale_color_manual(values = c("royalblue3", "indianred")) +
  scale_shape_manual(values=seq(0, 15)) +
  geom_text(aes(label = name), hjust = 0.5, vjust = 0.5) +
  stat_ellipse(aes(group = Treatment), type = "norm", level = 0.95) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_bw() + # Set the background to white
  theme(axis.text.x = element_text(size = 14, face="bold", color="black"),
        axis.text.y = element_text(size = 14, face="bold", color="black"),
        axis.title = element_text(size = 14, face="bold", color="black"),
        legend.title = element_text(size = 14, face="bold", color="black"),
        legend.text = element_text(size=14, face="bold", color="black"),
        #panel.grid.major = element_blank(), # Remove grid lines
        #panel.grid.minor = element_blank(),
        axis.line = element_line(), # Add axis lines
        axis.ticks = element_line()) +   # Add tick marks 
        theme(plot.title = element_text(face = "bold", hjust = 0.5)) # make title bold
  p + ggtitle("Ovaries Temperature vs Control PCA")



```


You may only have a list of differentially expressed genes, with ENSEMBL id's and not gene names. Use Biomart packages to resolve this.
```{r}
#read in your significant gene ids
ensembl.ids <- read.delim('sig_genes_female_all_group_age.csv', header =F)
head (ensembl.ids)

#biomart
listEnsembl()
ensembl <- useEnsembl(biomart = "genes")
datasets <- listDatasets(ensembl)
datasets

ensembl.con <- useMart("ensembl", dataset = "drerio_gene_ensembl")

attr <- listAttributes(ensembl.con)
filters <- listFilters(ensembl.con)

IDs <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype', 'external_gene_name'),
      filters = "ensembl_gene_id",
      values = ensembl.ids$V1,
      mart = ensembl.con)


head(IDs)
write.csv(IDs, file="genes_AMG_te.csv")



#load in your deseq2 output file, and add ensembl_id to first column header
resid <- read.csv('te_tocontrol_deseq2.csv')
# Merging two tables, syntax showed here are in full forms
annot.table <- merge(x = resid, y = IDs,
                     by.x = "ensembl_id", by.y =
                       "ensembl_gene_id", all.x = T, all.y = F )
# export your annotated results
write.csv(annot.table, file="annotated_te_tocontrol_deseq2.csv")


```


Making an Enhanced volcano plot
```{r}

# assign the results file name to a variable
deseq_results_file <- 'telescope_volcano.tsv'

# load data
deseq_results <-read_tsv(deseq_results_file,
           col_types = cols(Chr = col_character(),
                            Strand = col_character()))


head(deseq_results)
glimpse(deseq_results)
View(deseq_results)

EnhancedVolcano(
  deseq_results, # results data frame
  lab = deseq_results$gene,
  x = 'log2FoldChange', # column name of log2 fold change
  y = 'padj' # column name of adjusted pvalue
)


#to add the top 10 most sig de genes by p calue
genes_to_label <- deseq_results %>%
  arrange(padj) %>%#change log2fc to padj for by sig p value
  pull(miRNA) %>%
  head (30)

( ?EnhancedVolcano )

EnhancedVolcano(
  deseq_results, # results data frame
  lab = deseq_results$TE,
  x = 'log2FoldChange', # column name of log2 fold change
  y = 'padj', # column name of adjusted pvalue
  col = rocket(6), # Use Viridis color palette with 4 colors
  pCutoff = 50e-03,
  FCcutoff = log2(3.0),
  ylim = c(0, 3.5),
  xlim = c(-7, 7),
  min.segment.length = 0.1,
  labFace = "bold",
  labSize = 2.5,
  pointSize = 1.5,
  drawConnectors = TRUE,
  widthConnectors = 0.2,
  title = 'Enhanced Volcano',
  subtitle = 'RNA-seq Telescope:All adult samples, SEG v NEG'
)


```


Making a heatmap
```{r}
#########################################################
### B) Reading in data and transform it into matrix format
#########################################################

data <- read.csv("sig_de_mirnas_40_volcano.csv", comment.char="#")
rnames <- data[,1]                            # assign labels in column 1 to "rnames"
mat_data <- data.matrix(data[,2:ncol(data)])  # transform column 2-5 into a matrix
rownames(mat_data) <- rnames                  # assign row names

head(mat_data)
#########################################################
### C) Customizing and plotting the heat map
#########################################################

# creates a own color palette from red to green
my_palette <- colorRampPalette(c("red", "yellow", "green"))(n = 299)

# (optional) defines the color breaks manually for a "skewed" color transition
col_breaks = c(seq(-1,0,length=100),  # for red
               seq(0.01,0.8,length=100),           # for yellow
               seq(0.81,1,length=100))             # for green

#usethis::edit_r_environ("project") 

heatmap.2(mat_data,
          cellnote = mat_data,  # same data set for cell labels
          main = "Correlation", # heat map title
          notecol="black",      # change font color of cell labels to black
          density.info="none",  # turns off density plot inside color legend
          trace="none",         # turns off trace lines inside the heat map
          margins =c(12,9),     # widens margins around plot
          col=my_palette,       # use on color palette defined earlier
          breaks=col_breaks,    # enable color transition at specified limits
          dendrogram="row",     # only draw a row dendrogram
          Colv="NA")            # turn off column clustering


```

Making a Venn Diagram
```{r}
# Venn diagram

# Generate example data.
genes <- paste0("gene", 1:1000)
set.seed(20210701)
gene_list <- list(Female_mRNA = sample(genes, 41),
                  Male_mRNA = sample(genes, 3))

# Construct a Venn object
venn <- Venn(gene_list)
data <- process_data(venn, shape_id == "201")

p <-ggplot() +
  geom_sf(aes(fill = count, alpha = 0.5), data = venn_region(data)) +
  geom_sf(aes(color = id), data = venn_setedge(data), show.legend = FALSE) +
  geom_sf_text(aes(label = ifelse(name == "Female_mRNA", "Ovaries miRNAs", ifelse(name == "Male_mRNA", "Testes miRNAs", ""))), size = 5, data = venn_setlabel(data)) +
  geom_sf_text(aes(label = ifelse(name == "Female_mRNA", "41", ifelse(name == "Male_mRNA", "3", "0"))), size =12, fill = "transparent",  data = venn_region(data)) +
  
  theme_void()

#p + scale_color_brewer(palette = "Greys") # change outline
p + scale_color_brewer(palette = "Greys") +
    scale_fill_distiller(palette = "BuGn", direction =1) +
    theme(legend.position = "none") 


 
```

Plotting normalised TE count data 
```{r}

# Create a data frame from the image data
data <- data.frame(
  Treatment = c("Ctrl", "Ctrl", "Ctrl", "Ctrl", "Ctrl", "Ctrl", "Temp", "Temp", "Temp", "Temp", "Temp", "Temp"),
  gene = c("TC1", "TC2", "TC3", "TC4", "TC5", "TC6", "TT1", "TT2", "TT3", "TT4", "TT5", "TT6"),
  LINE = c(231832.5,	199312.1,	256774.5,	306471.7,	222973.7,	218331.7,	286619.1,	229882.8,	208081.1,	227034.4,	235012.6,	205713.1),
  LTR = c(700566.8,	586902.4,	667225.1,	844806.6,	588061.6,	633582.1,	788374.7,	617217,	562469.4,	636008.3,	659053,	640062.1),
  DNA = c(2445814,	2458397,	2197487,	2430212,	2137882,	2218152,	2388586,	2149667,	2126687,	2140268,	2149046,	2194718),
  SATELLITE = c(162961.1,	112603.5,	173118.6,	495027.2,	114121.9,	129002.9,	369314.5,	246892.6,	146294.2,	157543.5,	192203.7,	119720.4),
  SINE = c(38433.44,	28798.43,	37159.43,	45979.56,	24295.2,	32738.49,	44325.81,	29742.42,	28013.48,	30798.75,	30592.51,	28739.34),
  RC = c(292187.6,	264558.1,	294513.5,	339963.3,	244460.6,	265644.8,	326465,	265819,	253443.5,	272502.6,	276538.6,	266539.2)
)

# Reshape the data from wide to long format
data_long <- gather(data, key = Variable, value = Value, -Treatment, -gene)

# Create a box plot with data points and facets by variable
p <- ggplot(data_long, aes(x = Treatment, y = Value, fill = Treatment)) +
  geom_boxplot() +
  geom_point(size = 3, alpha = 0.5) +
  facet_grid(. ~ Variable, scales = "free_y") +
  ggtitle("RNASeq Telescope: Testes TE class counts") +
  xlab("Treatment") +
  ylab("Normalized Counts (Log10 Scale)") +
  scale_fill_manual(values = c("royalblue3", "indianred")) +
  scale_y_log10() +
  theme_bw() +
  theme(axis.text = element_text(size = 10, face = "bold", color = "black"),
        axis.title = element_text(size = 14, face = "bold", color = "black"),
        axis.line = element_line(),
        axis.ticks = element_line()) +
  theme(plot.title = element_text(face = "bold", hjust = 0.5)) +
  theme(legend.position = "none") 
  

p

p + scale_y_log10(limits = c(3e+04, 3e+06)) # just to give a bit more space to manually add significance stars on the plot later


# Calculate paired t-tests
ttest_results <- data_long %>%
  group_by(Variable) %>%
  summarize(p_value = t.test(Value ~ Treatment, paired = TRUE, alternative = "greater")$p.value)


```

Calculating normalised counts with DESeq2 normalised counts outputs
```{r}

# Read the CSV file
data <- read.csv("male_telescope_normalised_counts.csv", header = TRUE)
head(data)

# Find and print rows containing "hAT", "TDR", and "DNA"
filtered_data <- data[grep("Tc1|RC|Mariner", data$X, ignore.case = TRUE), ]
cat("Rows with 'Tc1', 'RC' or 'Mariner':\n")
print(filtered_data)

# Count and print the sum of each column (columns 2-13)
column_sums <- colSums(filtered_data[, 2:13])
cat("Sum of each column (columns 2-13):\n")
print(column_sums)

```


